### PR TITLE
log,logstream: structured log emission & consumption mechanism

### PR DESCRIPTION
**Pull request by abarganier**
**Note: please only consider the final two commits. The first two commits are being reviewed separately in https://github.com/cockroachdb/cockroach/pull/119416**

---

This PR introduces `log.Structured` to the log package API. It aims to
serve as a prototype for the log facility we will use for exporting
"exhaust" from CRDB in the form of JSON objects. The intention is that
this exhaust can be consumed externally and be sufficient enough to
build features around.

This iteration has some serious limitations, the main one being that it
is not redaction-aware. The `log.StructuredEvent` exists alongside it
for now. Both implementations are quite similar, so they should probably
be reconciled and/or combined, but this is left as a TODO to avoid
slowing down the prototyping process. For now, it's sufficient for
prototyping.

The patch also introduces a new logging channel explicitly for the new
`log.Structured` API, called `STRUCTURED_EVENTS`. The ability to segment these
types of logs from the rest of the logs is what motivates this separate
channel. The name works for now, but we should consider if there's a
better name available.

The PR also expands upon the new structured logging facilities, adding a
mechanism to consume emitted structured logs internally.

This is done primarily via the new pkg/obs/logstream package, which
handles buffering, routing, & processing of events logged via
log.Structured.

It can be used in conjunction with the eventagg package, and the
KVProcessor interface, to provide users of the eventagg package a way
to consume streams of events flushed from their aggregations. This
enables engineers to use the aggregated data flushed from their
aggregations to build features internal to CRDB. Features that are
powered by the same data that could be consumed externally via the
STRUCTURED_EVENTS log channel.

The provided log config can be updated to make use of this new channel.
For example:
```
sinks:
  file-groups:
    structured-events:
      channels: [STRUCTURED_EVENTS]
```

The changes aim to complete the eventagg pipeline/ecosystem, which now
allows engineers to use common facilities to define aggregations, log
the aggregated results, and consume the logged events internally as
input data.

Finally, it completes the toy StmtStats example by defining a processor
for the aggregated events that are logged.

The below diagram outlines the end-to-end architecture of the system:

Screenshot 2024-04-26 at 3 59 21 PM

Release note: none

Epic: CRDB-35919

### CODE REVIEW

**Comment by kyle-a-wong** [comment]
@abarganier I rebased on top of origin master so only the structured log related changes are in this PR

**Review by kyle-a-wong — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124058)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)

___
*[`pkg/obs/eventagg/kv_processor.go` line 0 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/124058#-NzTQ4jM7X9C7El3KbCS:-NzTQ4jM7X9C7El3KbCT:b-id13s) ([raw file](https://github.com/cockroachdb/cockroach/blob/236f2f83d3dd19a4e971bb18936b5b4a14d22890/pkg/obs/eventagg/kv_processor.go#L0)):*
It seems like this may be unnecessary based on the commit you and I worked on for the job history logging. Specifically in this snippet of code where we define `StructuredLogProcessor` https://github.com/cockroachdb/cockroach/pull/124242/commits/9b61fa7da1e83ab54dfcaf8cee50ffc24253047f#diff-e49dbd78ca75b049dbcc24c3d7441af89c7e497ecf7b5b48c96367885e6f2896R29

Then, in `func processEmittedStmtStats` we can just change the signature to 
```
func processEmittedStmtStats(
	_ context.Context,
	kvl KeyValueLog
)
```

I think this would also require changes to KeyValueLog to be generic. 
___
*[`pkg/obs/logstream/processor.go` line 20 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/124058#-NzTP3ab2P_MbOOlN-K1:-NzTP3ab2P_MbOOlN-K2:b31uvwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/236f2f83d3dd19a4e971bb18936b5b4a14d22890/pkg/obs/logstream/processor.go#L20)):*
> ```Go
> //
> // See RegisterProcessor for more details on how to create & register a processor with this package.
> type Processor interface {
> ```

Is there a reason you prefer Process to  use `any` as an argument type instead of using generics, like:
```
type Processor[T any] interface {
	// Process processes a single log stream event.
	Process(context.Context, t) error

```



_Code quote:_
```Go
ype Processor interface 
```

**Review by dhartunian — APPROVED** [review]
LGTM 🎉

File: pkg/obs/logstream/async_processor_router.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+	bs.mu.Lock()
+	defer bs.mu.Unlock()
+	// Append the event to the buffer.
+	err := bs.mu.buf.appendEvent(e)
+	if err != nil {
+		return err
+	}
+
+	flush := bs.triggerLen > 0 && bs.mu.buf.len() >= bs.triggerLen
+	if flush {
```
[the remark this answers is not in the document]
Comment (abarganier) [reply pkg/obs/logstream/async_processor_router.go:154]: Good idea! Changed to `maybeFlushAsyncLocked` and encapsulated all the decision making there.

File: pkg/util/log/structured_v2.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	STATEMENT_STATS EventType = "stmt_stats"
+)
+
+var (
+	trimPrefix = []byte("{")
+	trimSuffix = []byte("}")
+)
+
+// StructuredMeta is a metadata object that accompanies each event emitted via Structured.
+type StructuredMeta struct {
```
[the remark this answers is not in the document]
Comment (abarganier) [reply pkg/util/log/structured_v2.go]: Went ahead and just implemented it, it wasn't too much effort.

**Review by abarganier — COMMENTED** [review]
TFTR! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124058#-:-NzeDdkcDEy43huOji70:bz5m8k4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @kyle-a-wong)

___
*[`pkg/obs/eventagg/kv_processor.go` line  at r6](https://reviewable.io/reviews/cockroachdb/cockroach/124058#-NzTQ4jM7X9C7El3KbCS:-NzeDI4G595fFiLk9Ak2:ba3q0if) ([raw file](https://github.com/cockroachdb/cockroach/blob/236f2f83d3dd19a4e971bb18936b5b4a14d22890/pkg/obs/eventagg/kv_processor.go#L)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

It seems like this may be unnecessary based on the commit you and I worked on for the job history logging. Specifically in this snippet of code where we define `StructuredLogProcessor` https://github.com/cockroachdb/cockroach/pull/124242/commits/9b61fa7da1e83ab54dfcaf8cee50ffc24253047f#diff-e49dbd78ca75b049dbcc24c3d7441af89c7e497ecf7b5b48c96367885e6f2896R29

Then, in `func processEmittedStmtStats` we can just change the signature to 
```
func processEmittedStmtStats(
	_ context.Context,
	kvl KeyValueLog
)
```

I think this would also require changes to KeyValueLog to be generic.
</blockquote></details>

Since the interfaces have changed, this is no longer relevant. However, making `KeyValueLog` generic was an idea we ran with which made things much more convenient when it comes to making type assertions!
___
*[`pkg/obs/logstream/processor.go` line 20 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/124058#-NzTP3ab2P_MbOOlN-K1:-NzeDROg-bZBraqchulE:behy1ll) ([raw file](https://github.com/cockroachdb/cockroach/blob/236f2f83d3dd19a4e971bb18936b5b4a14d22890/pkg/obs/logstream/processor.go#L20)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

Is there a reason you prefer Process to  use `any` as an argument type instead of using generics, like:
```
type Processor[T any] interface {
	// Process processes a single log stream event.
	Process(context.Context, t) error

```
</blockquote></details>

As discussed on Slack, using generics posed some weird challenges given some limitations we found with Golang. The added cruft required to make it work didn't seem worth it, so we're sticking with the `any` interface for simplicity.

**Review by kyle-a-wong — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124058#-:-NzeHDgkEiL_bWyKkdrR:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dhartunian)

**Comment by abarganier** [comment]
TFTRs! 

bors r+

**Pull request merged**